src/test/java/com/proofpoint/discovery/TestIpHostnameAuthManager.java

```diff
@@ -0,0 +1,242 @@
+    }
+
+    @Test
+    public void testDeleteMismatch()
+    {
+        Id<Node> nodeId = Id.random();
+        DynamicAnnouncement toDelete = DynamicAnnouncement.copyOf(new DynamicAnnouncement("testing", "general", "/location", ImmutableSet.of(
+                new DynamicServiceAnnouncement(Id.random(), "storage", ImmutableMap.of("http", "http://localhost:4111"))
+        ))).setAnnouncer("10.20.30.40").build();
+        dynamicStore.put(nodeId, toDelete);
+
+        AuthManager authManager = new IpHostnameAuthManager(dynamicStore, configStore);
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        when(request.getRemoteAddr()).thenReturn("127.0.0.1");
+        assertThatExceptionOfType(ForbiddenException.class)
+                .isThrownBy(() -> authManager.checkAuthDelete(nodeId, request))
+                .withMessageContaining("HTTP 403")
+                .withNoCause();
+    }
+
+    @Test
+    public void testDeleteInvalidIp()
+    {
+        Id<Node> nodeId = Id.random();
+        DynamicAnnouncement toDelete = DynamicAnnouncement.copyOf(new DynamicAnnouncement("testing", "general", "/location", ImmutableSet.of(
+                new DynamicServiceAnnouncement(Id.random(), "storage", ImmutableMap.of("http", "http://localhost:4111"))
+        ))).setAnnouncer("10.20.30.40").build();
+        dynamicStore.put(nodeId, toDelete);
+
+        AuthManager authManager = new IpHostnameAuthManager(dynamicStore, configStore);
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        when(request.getRemoteAddr()).thenReturn("10.10.10.10.10");
+        assertThatExceptionOfType(ForbiddenException.class)
+                .isThrownBy(() -> authManager.checkAuthDelete(nodeId, request))
+                .withMessageContaining("HTTP 403")
+                .withNoCause();
+    }
+
+    @Test
+    public void testAnnounce()
+    {
+        Id<Node> nodeId = Id.random();
+        DynamicServiceAnnouncement localhostServiceAnnouncement = new DynamicServiceAnnouncement(Id.random(), "storage", ImmutableMap.of("http", "http://localhost:4111"));
+        DynamicAnnouncement localhostAnnouncement = new DynamicAnnouncement("testing", "general", "/location", ImmutableSet.of(localhostServiceAnnouncement));
+        AuthManager authManager = new IpHostnameAuthManager(dynamicStore, configStore);
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        when(request.getRemoteAddr()).thenReturn("127.0.0.1");
+        //Announce new
+        assertThatCode(() -> authManager.checkAuthAnnounce(nodeId, localhostAnnouncement, request)).doesNotThrowAnyException();
+        dynamicStore.put(nodeId, DynamicAnnouncement.copyOf(localhostAnnouncement).setAnnouncer("127.0.0.1").build());
+        //Re-announce
+        assertThatCode(() -> authManager.checkAuthAnnounce(nodeId, localhostAnnouncement, request)).doesNotThrowAnyException();
```

Comment:
Test announce when entry has null announcer?
Test announce when entry is a tombstone?

Reply:
The first is resolved. I don't see a test for when the entry is a tombstone.